vespalib/src/vespa/vespalib/stllike/cache.hpp

```diff
@@ -7,55 +7,224 @@
 
 namespace vespalib {
 
-template< typename P >
-cache<P> &
+template <typename P>
+cache<P>::SizeConstrainedLru::SizeConstrainedLru(cache& owner, size_t capacity_bytes)
+    : Lru(Lru::UNLIMITED),
+      _owner(owner),
+      _size_bytes(0),
+      _capacity_bytes(capacity_bytes)
+{
+}
+
+template <typename P>
+cache<P>::SizeConstrainedLru::~SizeConstrainedLru() = default;
+
+template <typename P>
+bool
+cache<P>::SizeConstrainedLru::removeOldest(const typename P::value_type& kv) {
+    const size_t sz_now = size_bytes();
+    // TODO shouldn't this be size > capacity, not >= ? Not touching it for now...
+    bool remove(Lru::removeOldest(kv) || (sz_now >= capacity_bytes()));
+    if (remove) {
+        sub_size_bytes(_owner.calcSize(kv.first, kv.second._value));
+    }
+    return remove;
+}
+
+template <typename P>
+bool
+cache<P>::SizeConstrainedLru::has_key(const KeyT& key) const noexcept {
+    return Lru::hasKey(key);
+}
+
+template <typename P>
+void
+cache<P>::SizeConstrainedLru::insert_and_update_size(const KeyT& key, ValueT value) {
+    const auto kv_size = _owner.calcSize(key, value);
+    auto insert_res = Lru::insert(key, std::move(value));
+    assert(insert_res.second);
+    // Must be updated _after_ insert, since the underlying eviction policy (removeOldest)
+    // transitively consults this value as part of the insertion itself.
+    add_size_bytes(kv_size);
```

Comment:
Fixed. This was done because the existing tests expected it. Have updated size tracking behavior and the associated tests. Soft limit now much less soft.